File: 07-04-network-image-viewmodel/java-07-04-network-image-viewmodel/app/src/main/java/com/example/viewModelImage/Repository.java
package com.example.viewModelImage;

import com.google.gson.FieldNamingPolicy;
import com.google.gson.Gson;
import com.google.gson.GsonBuilder;

import retrofit2.Call;
import retrofit2.Retrofit;
import retrofit2.converter.gson.GsonConverterFactory;

public class Repository {

    private final PokeAPI pokeAPI;
    private static Repository instance = new Repository();

    private Repository() {
        Gson gson = new GsonBuilder()
                .setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES)
                .create();
        Retrofit retrofit = new Retrofit.Builder()
                .baseUrl("https://pokeapi.co/api/v2/")
                .addConverterFactory(GsonConverterFactory.create(gson))
                .build();
        pokeAPI = retrofit.create(PokeAPI.class);
    }

    public static Repository getInstance() {
        return instance;
    }

    Call<Response> listPokemons() {
        return pokeAPI.listPokemons();
    }

    Call<Response> listPokemons(String offset, String limit) {
        return pokeAPI.listPokemons(offset, limit);
    }

    Call<PokemonResponse> getPokemon(int pid) {
        return pokeAPI.getPokemon(pid);
    }
}

File: 08-02-network-image-coroutine/app/src/main/java/com/example/viewModelImageCoroutine/Repository.kt
package com.example.viewModelImageCoroutine

import com.google.gson.FieldNamingPolicy
import com.google.gson.Gson
import com.google.gson.GsonBuilder
import retrofit2.Retrofit
import retrofit2.converter.gson.GsonConverterFactory

object Repository {
    private val pokeAPI: PokeAPI

    init {
        val gson: Gson = GsonBuilder()
                .setFieldNamingPolicy(FieldNamingPolicy.LOWER_CASE_WITH_UNDERSCORES)
                .create()
        val retrofit: Retrofit = retrofit2.Retrofit.Builder()
                .baseUrl("https://pokeapi.co/api/v2/")
                .addConverterFactory(GsonConverterFactory.create(gson))
                .build()
        pokeAPI = retrofit.create(PokeAPI::class.java)
    }

    fun listPokemons(): retrofit2.Call<Response> {
        return pokeAPI.listPokemons()
    }

    fun listPokemons(offset: String?, limit: String?): retrofit2.Call<Response> {
        return pokeAPI.listPokemons(offset!!, limit!!)
    }

    suspend fun getPokemon(pid: Int): PokemonResponse {
        return pokeAPI.getPokemon(pid)
    }
}